File: 2022/day14/php/solution.php
#!/usr/bin/env php
<?php

/**
 * Day 14: Regolith Reservoir - Falling sand simulation
 */

/**
 * Parse rock paths and return array of rock positions.
 */
function parsePaths(string $text): array {
    $rocks = [];
    $lines = explode("\n", trim($text));

    foreach ($lines as $line) {
        $points = explode(' -> ', $line);
        for ($i = 0; $i < count($points) - 1; $i++) {
            list($x1, $y1) = array_map('intval', explode(',', $points[$i]));
            list($x2, $y2) = array_map('intval', explode(',', $points[$i + 1]));

            // Draw line from (x1, y1) to (x2, y2)
            if ($x1 === $x2) {
                // Vertical line
                for ($y = min($y1, $y2); $y <= max($y1, $y2); $y++) {
                    $rocks["$x1,$y"] = true;
                }
            } else {
                // Horizontal line
                for ($x = min($x1, $x2); $x <= max($x1, $x2); $x++) {
                    $rocks["$x,$y1"] = true;
                }
            }
        }
    }

    return $rocks;
}

/**
 * Simulate one unit of sand falling.
 * Returns the resting position as "x,y", or null if sand falls into abyss.
 */
function simulateSand(array &$blocked, int $maxY, bool $floor = false): ?string {
    $x = 500;
    $y = 0;

    while (true) {
        // Check if sand has fallen below all rocks (into abyss)
        if (!$floor && $y > $maxY) {
            return null;
        }

        // Try to move down
        if ($floor && $y + 1 === $maxY + 2) {
            // Hit the floor
            return "$x,$y";
        } elseif (!isset($blocked["$x," . ($y + 1)])) {
            $y++;
        }
        // Try to move down-left
        elseif (!isset($blocked[($x - 1) . "," . ($y + 1)])) {
            $x--;
            $y++;
        }
        // Try to move down-right
        elseif (!isset($blocked[($x + 1) . "," . ($y + 1)])) {
            $x++;
            $y++;
        }
        // Sand comes to rest
        else {
            return "$x,$y";
        }
    }
}

/**
 * Part 1: Count sand units that come to rest before sand falls into abyss.
 */
function part1(string $text): int {
    $rocks = parsePaths($text);

    // Find max Y coordinate
    $maxY = 0;
    foreach (array_keys($rocks) as $key) {
        list($x, $y) = explode(',', $key);
        $maxY = max($maxY, (int)$y);
    }

    $blocked = $rocks;
    $count = 0;

    while (true) {
        $pos = simulateSand($blocked, $maxY);
        if ($pos === null) {
            break;
        }
        $blocked[$pos] = true;
        $count++;
    }

    return $count;
}

/**
 * Part 2: Count sand units until source is blocked (with floor).
 */
function part2(string $text): int {
    $rocks = parsePaths($text);

    // Find max Y coordinate
    $maxY = 0;
    foreach (array_keys($rocks) as $key) {
        list($x, $y) = explode(',', $key);
        $maxY = max($maxY, (int)$y);
    }

    $blocked = $rocks;
    $count = 0;

    while (true) {
        $pos = simulateSand($blocked, $maxY, true);
        $blocked[$pos] = true;
        $count++;
        if ($pos === "500,0") {
            break;
        }
    }

    return $count;
}

// Main execution
$scriptDir = dirname(__FILE__);
$inputFile = $scriptDir . '/../input.txt';
$text = file_get_contents($inputFile);

echo "Part 1: " . part1($text) . "\n";
echo "Part 2: " . part2($text) . "\n";

File: 2022/day14/ruby/solution.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'set'

def parse_paths(text)
  rocks = Set.new
  text.strip.split("\n").each do |line|
    points = line.split(' -> ')
    (0...points.length - 1).each do |i|
      x1, y1 = points[i].split(',').map(&:to_i)
      x2, y2 = points[i + 1].split(',').map(&:to_i)

      if x1 == x2 # vertical line
        ([y1, y2].min..[y1, y2].max).each { |y| rocks.add([x1, y]) }
      else # horizontal line
        ([x1, x2].min..[x1, x2].max).each { |x| rocks.add([x, y1]) }
      end
    end
  end
  rocks
end

def simulate_sand(blocked, max_y, floor: false)
  x, y = 500, 0

  loop do
    # Check if sand has fallen below all rocks (into abyss)
    return nil if !floor && y > max_y

    # Try to move down
    if floor && y + 1 == max_y + 2
      # Hit the floor
      return [x, y]
    elsif !blocked.include?([x, y + 1])
      y += 1
    # Try to move down-left
    elsif !blocked.include?([x - 1, y + 1])
      x -= 1
      y += 1
    # Try to move down-right
    elsif !blocked.include?([x + 1, y + 1])
      x += 1
      y += 1
    # Sand comes to rest
    else
      return [x, y]
    end
  end
end

def part1(text)
  rocks = parse_paths(text)
  max_y = rocks.map { |_, y| y }.max
  blocked = rocks.dup
  count = 0

  loop do
    pos = simulate_sand(blocked, max_y)
    break if pos.nil?

    blocked.add(pos)
    count += 1
  end

  count
end

def part2(text)
  rocks = parse_paths(text)
  max_y = rocks.map { |_, y| y }.max
  blocked = rocks.dup
  count = 0

  loop do
    pos = simulate_sand(blocked, max_y, floor: true)
    blocked.add(pos)
    count += 1
    break if pos == [500, 0]
  end

  count
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')
  text = File.read(input_file)

  puts "Part 1: #{part1(text)}"
  puts "Part 2: #{part2(text)}"
end

main if __FILE__ == $PROGRAM_NAME
